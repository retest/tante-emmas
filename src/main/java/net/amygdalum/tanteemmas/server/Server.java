package net.amygdalum.tanteemmas.server;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpHeaders;
import io.vertx.core.http.HttpServer;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.impl.StaticHandlerImpl;
import io.vertx.ext.web.templ.HandlebarsTemplateEngine;
import io.vertx.ext.web.templ.TemplateEngine;
import net.amygdalum.tanteemmas.external.SimulatedDateSource;
import net.amygdalum.tanteemmas.external.SimulatedDaytimeSource;
import net.amygdalum.tanteemmas.external.SimulatedTimeProvider;
import net.amygdalum.tanteemmas.external.SimulatedWeatherSource;
import net.amygdalum.tanteemmas.sources.DateSource;
import net.amygdalum.tanteemmas.sources.DaytimeSource;
import net.amygdalum.tanteemmas.sources.TimeProvider;
import net.amygdalum.tanteemmas.sources.WeatherSource;

public class Server extends AbstractVerticle {

	private final Vertx vertx;

	private final CustomerRepo customers;
	private final ProductRepo products;
	private final TemplateEngine engine;
	private final List<Map<String, Object>> orders = new ArrayList<>();

	private TimeProvider time;
	private DateSource date;
	private DaytimeSource daytime;
	private WeatherSource weather;

	public Server() {
		engine = HandlebarsTemplateEngine.create().setExtension("html");
		products = new ProductRepo().init();
		customers = new CustomerRepo().init();
		time = new SimulatedTimeProvider();
		date = new SimulatedDateSource(time);
		daytime = new SimulatedDaytimeSource(time);
		weather = new SimulatedWeatherSource(time, date);

		vertx = Vertx.vertx();
	}

	@Override
	public void start() {
		Router router = Router.router(vertx);
		router.route("/speed/:speed").handler(this::speed);
		router.route("/login").handler(this::login);
		router.route("/logout").handler(this::logout);
		router.route("/order/:product").handler(this::order);
		router.route("/prices").handler(this::prices);
		router.route("/showPrices").handler(this::showPrices);
		router.route("/showLogin").handler(this::showLogin);

		router.route("/css/*").handler(new StaticHandlerImpl("src/main/resources/css", null));
		router.route("/fonts/*").handler(new StaticHandlerImpl("src/main/resources/fonts", null));
		router.route("/img/*").handler(new StaticHandlerImpl("src/main/resources/img", null));
		router.route("/js/*").handler(new StaticHandlerImpl("src/main/resources/js", null));

		router.route().handler(this::show);

		HttpServer server = vertx.createHttpServer();
		server.requestHandler(router::accept).listen(8080);
		System.out.println("Started tante-emma-server.");
	}

	@Override
	public void stop() throws Exception {
		System.out.println("Stopped tante-emma-server.");
	}

	public void speed(RoutingContext context) {
		long speed = Long.parseLong(context.request().getParam("speed"));
		time.setSpeed(speed);
		context.reroute("/prices");
	}

	public void login(RoutingContext context) {
		String name = context.request().getParam("customer");
		Customer customer = customers.getCustomer(name);
		PriceCalculator.customer = customer;
		context.reroute("/prices");
	}

	public void logout(RoutingContext context) {
		PriceCalculator.customer = null;
		context.reroute("/showLogin");
	}

	public void order(RoutingContext context) {
		if (PriceCalculator.customer == null) {
			context.next();
			return;
		}
		PriceCalculator prices = new PriceCalculator(date, daytime, weather);
		String name = context.request().getParam("product");
		Map<String, Object> product = new HashMap<>(products.getProduct(name));
		product.put("price", prices.computePrice(product));
		orders.add(product);
		context.data().put("orders", orders);
		context.reroute("/prices");
	}

	public void prices(RoutingContext context) {
		if (PriceCalculator.customer == null) {
			context.next();
			return;
		}
		try {
			PriceCalculator prices = new PriceCalculator(date, daytime, weather);
			List<Map<String, Object>> priceTable = new ArrayList<>();
			context.data().put("products", priceTable);
			for (Map<String, Object> product : products.getProducts()) {
				BigDecimal price = prices.computePrice(product);

				Map<String, Object> productWithPrice = new HashMap<>(product);
				productWithPrice.put("price", price);
				priceTable.add(productWithPrice);
			}
			context.next();
		} catch (RuntimeException e) {
			throw e;
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	public void show(RoutingContext context) {
		context.data().put("date", date.getDate());
		context.data().put("daytime", daytime.getDaytime());
		context.data().put("weather", weather.getWeather());

		if (PriceCalculator.customer == null) {
			context.reroute("/showLogin");
		} else {
			context.reroute("/showPrices");
		}
	}

	public void showPrices(RoutingContext context) {
		context.data().put("speed", time.getSpeed());
		context.data().put("incspeed", time.getSpeed() * 10);
		context.data().put("decspeed", time.getSpeed() / 10);
		context.data().put("user", PriceCalculator.customer.name);

		engine.render(context, "src/main/resources/index.html", res -> {
			if (res.succeeded()) {
				context.response().putHeader(HttpHeaders.CONTENT_TYPE, "text/html").end(res.result());
			} else {
				context.fail(res.cause());
			}
		});
	}

	public void showLogin(RoutingContext context) {
		engine.render(context, "src/main/resources/login.html", res -> {
			if (res.succeeded()) {
				context.response().putHeader(HttpHeaders.CONTENT_TYPE, "text/html").end(res.result());
			} else {
				context.fail(res.cause());
			}
		});
	}

	public static void main(String[] args) {
		new Server().startServer();
	}

	public void startServer() {
		vertx.deployVerticle(Server.class.getName());
	}

	public void stopServer() {
		vertx.close();
	}

	public void stopServerAndWait() throws InterruptedException {
		final CountDownLatch latch = new CountDownLatch(1);
		vertx.close( e-> latch.countDown() );
		latch.await();
	}
}
