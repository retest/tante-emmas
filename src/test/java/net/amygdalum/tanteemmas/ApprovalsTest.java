package net.amygdalum.tanteemmas;

import org.approvaltests.Approvals;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

import net.amygdalum.tanteemmas.server.Server;
import net.amygdalum.tanteemmas.testutils.ChromeDriverFactory;

public class ApprovalsTest {

	private Server server;
	private WebDriver driver;

	@Before
	public void setup() {
		server = new Server();
		server.startServer();

		driver = ChromeDriverFactory.createNewInstance();
	}

	@Test
	public void check_order() throws Exception {
		driver.get("http://localhost:8080/logout");
		// login
		driver.findElement(By.name("customer")).sendKeys("Max");
		driver.findElement(By.name("login")).click();

		// test
		driver.findElement(By.xpath("//tr[4]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[4]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[2]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[3]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[2]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[7]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[6]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[5]/td[3]/a")).click();
		driver.findElement(By.xpath("//tr[3]/td[3]/a")).click();

		Approvals.verifyHtml(driver.getPageSource());
	}

	@After
	public void tearDown() throws InterruptedException {
		driver.quit();
		server.stopServerAndWait();
	}
}
