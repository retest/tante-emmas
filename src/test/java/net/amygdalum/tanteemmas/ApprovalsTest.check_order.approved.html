<!DOCTYPE html><html xmlns="http://www.w3.org/1999/xhtml" lang="en" class="csstransforms csstransforms3d csstransitions"><head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>eNno Bootstrap Template</title>

    <!-- Bootstrap -->
    <link href="/css/bootstrap.min.css" rel="stylesheet" />
	<link rel="stylesheet" href="/css/animate.css" />
	<link rel="stylesheet" href="/css/font-awesome.min.css" />
	<link rel="stylesheet" href="/css/jquery.bxslider.css" />
	<link rel="stylesheet" type="text/css" href="/css/normalize.css" />
	<link rel="stylesheet" type="text/css" href="/css/demo.css" />
	<link rel="stylesheet" type="text/css" href="/css/set1.css" />
	<link href="/css/overwrite.css" rel="stylesheet" />
	<link href="/css/style.css" rel="stylesheet" />
    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
  </head>
  <body>
	<nav class="navbar navbar-default navbar-fixed-top" role="navigation">
		<div class="container">
			<!-- Brand and toggle get grouped for better mobile display -->
			<div class="navbar-header">
				<button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target=".navbar-collapse.collapse">
					<span class="sr-only">Toggle navigation</span>
					<span class="icon-bar"></span>
					<span class="icon-bar"></span>
					<span class="icon-bar"></span>
				</button>
				<a class="navbar-brand" href="index.html"><span>eNno</span></a>
			</div>
			<div class="navbar-collapse collapse">							
				<div class="menu">
					<ul class="nav nav-tabs" role="tablist">
						<li role="presentation" class="active"><a href="index.html">Home</a></li>
						<li role="presentation"><a href="services.html">Services</a></li>
						<li role="presentation"><a href="blog.html">Blog</a></li>
						<li role="presentation"><a href="portfolio.html">Portfolio</a></li>
						<li role="presentation"><a href="contact.html">Contact</a></li>						
					</ul>
				</div>
			</div>			
		</div>
	</nav>
	
	
	<div class="container">
		<div class="row">
			<div class="slider">
				<div class="img-responsive">
					<div class="bx-wrapper" style="max-width: 100%;"><div class="bx-viewport" style="width: 100%; overflow: hidden; position: relative; height: 360px;"><ul class="bxslider" style="width: 515%; position: relative; transition-duration: 0s; transform: translate3d(-1250px, 0px, 0px);"><li style="float: left; list-style: none; position: relative; width: 1170px;" class="bx-clone"><img src="/img/3.jpg" alt="" /></li>				
						<li style="float: left; list-style: none; position: relative; width: 1170px;"><img src="/img/1.jpg" alt="" /></li>								
						<li style="float: left; list-style: none; position: relative; width: 1170px;"><img src="/img/2.jpg" alt="" /></li>	
						<li style="float: left; list-style: none; position: relative; width: 1170px;"><img src="/img/3.jpg" alt="" /></li>			
					<li style="float: left; list-style: none; position: relative; width: 1170px;" class="bx-clone"><img src="/img/1.jpg" alt="" /></li></ul></div><div class="bx-controls bx-has-pager bx-has-controls-direction"><div class="bx-pager bx-default-pager"><div class="bx-pager-item"><a href="" data-slide-index="0" class="bx-pager-link active">1</a></div><div class="bx-pager-item"><a href="" data-slide-index="1" class="bx-pager-link">2</a></div><div class="bx-pager-item"><a href="" data-slide-index="2" class="bx-pager-link">3</a></div></div><div class="bx-controls-direction"><a class="bx-prev" href="">Prev</a><a class="bx-next" href="">Next</a></div></div></div>
				</div>	
			</div>
		</div>
	</div>
		
	<div class="container">
		<div class="row">
			<div class="col-md-6 col-md-offset-3">
				<div class="text-center">


	<h1>Tante Emmas Online Shop</h1>
	<h2>
		Logged in as: Max <a href="/logout">logout</a>
	</h2>


				</div>
				<hr />
			</div>
		</div>
	</div>
	
	<div class="container">
		<div class="row">
			<div class="box">
				<div class="col-md-4">

	<h1>Products</h1>
	<table>
		<tbody><tr>
		<th>Product</th>
		<th>Price</th>
		</tr> 
		
		<tr>
			<td>rain cape</td>
			<td>15.15</td>
			<td><a href="/order/rain cape">Order</a></td>
		</tr>
		
		<tr>
			<td>bathing suit</td>
			<td>25.25</td>
			<td><a href="/order/bathing suit">Order</a></td>
		</tr>
		
		<tr>
			<td>sunscreen</td>
			<td>2.02</td>
			<td><a href="/order/sunscreen">Order</a></td>
		</tr>
		
		<tr>
			<td>vodka</td>
			<td>9.90</td>
			<td><a href="/order/vodka">Order</a></td>
		</tr>
		
		<tr>
			<td>flour</td>
			<td>0.70</td>
			<td><a href="/order/flour">Order</a></td>
		</tr>
		
		<tr>
			<td>10 eggs</td>
			<td>2.62</td>
			<td><a href="/order/10 eggs">Order</a></td>
		</tr>
		
	</tbody></table>

				</div>
				
				<div class="col-md-4">

	<h1>Cart</h1>
	<table>
	  <tbody><tr>
		<th>Product</th>
		<th>Price</th>
		</tr> 
		
		<tr>
			<td>sunscreen</td>
			<td>2.02</td>
		</tr>
		
		<tr>
			<td>sunscreen</td>
			<td>2.02</td>
		</tr>
		
		<tr>
			<td>rain cape</td>
			<td>15.15</td>
		</tr>
		
		<tr>
			<td>bathing suit</td>
			<td>25.25</td>
		</tr>
		
		<tr>
			<td>rain cape</td>
			<td>15.15</td>
		</tr>
		
		<tr>
			<td>10 eggs</td>
			<td>2.62</td>
		</tr>
		
		<tr>
			<td>flour</td>
			<td>0.70</td>
		</tr>
		
		<tr>
			<td>vodka</td>
			<td>9.90</td>
		</tr>
		
		<tr>
			<td>bathing suit</td>
			<td>25.25</td>
		</tr>
		
	</tbody></table>

				</div>
				
				<div class="col-md-4">

	<h2>The day of the week</h2>
			<table>
				<tbody><tr>
					<td>Day of the week</td>
					<td>MONDAY</td>
				</tr>
		
				<tr>
					<td>Season</td>
					<td>WINTER</td>
				</tr>
		
				<tr>
					<td>Time of the day</td>
					<td>NIGHT</td>
				</tr>
		
				<tr>
					<td>Wind</td>
					<td>WIND</td>
				</tr>
		
				<tr>
					<td>Precipitation</td>
					<td>MIST</td>
				</tr>
		
				<tr>
					<td>Temperature</td>
					<td>COLD</td>
				</tr>
			</tbody></table>

				</div>
			</div>
		</div>
	</div>
	
	<footer>
		<div class="last-div">
			<div class="container">
				<div class="row">
					<div class="copyright">
						Â© 2014 eNno Multi-purpose theme | <a target="_blank" href="http://bootstraptaste.com">Bootstraptaste</a>
					</div>	
                    <!-- 
                        All links in the footer should remain intact. 
                        Licenseing information is available at: http://bootstraptaste.com/license/
                        You can buy this theme without footer links online at: http://bootstraptaste.com/buy/?theme=eNno
                    -->				
				</div>
			</div>
			<div class="container">
				<div class="row">
					<ul class="social-network">
						<li><a href="#" data-placement="top" title="Facebook"><i class="fa fa-facebook fa-1x"></i></a></li>
						<li><a href="#" data-placement="top" title="Twitter"><i class="fa fa-twitter fa-1x"></i></a></li>
						<li><a href="#" data-placement="top" title="Linkedin"><i class="fa fa-linkedin fa-1x"></i></a></li>
						<li><a href="#" data-placement="top" title="Pinterest"><i class="fa fa-pinterest fa-1x"></i></a></li>
						<li><a href="#" data-placement="top" title="Google plus"><i class="fa fa-google-plus fa-1x"></i></a></li>
					</ul>
				</div>
			</div>
			
			<a href="" class="scrollup"><i class="fa fa-chevron-up"></i></a>	
				
			
		</div>	
	</footer>
	
    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="/js/jquery-2.1.1.min.js"></script>
    <!-- Include all compiled plugins (below), or include individual files as needed -->
    <script src="/js/bootstrap.min.js"></script>
	<script src="/js/wow.min.js"></script>
	<script src="/js/jquery.easing.1.3.js"></script>
	<script src="/js/jquery.isotope.min.js"></script>
	<script src="/js/jquery.bxslider.min.js"></script>
	<script type="text/javascript" src="/js/fliplightbox.min.js"></script>
	<script src="/js/functions.js"></script>	
	<script type="text/javascript">$('.portfolio').flipLightBox()</script>
  
</body></html>
